core/command_queue_mt.cpp

```diff
@@ -76,6 +76,30 @@ CommandQueueMT::SyncSemaphore *CommandQueueMT::_alloc_sync_sem() {
 	return &sync_sems[idx];
 }
 
+bool CommandQueueMT::dealloc_one() {
+tryagain:
+	if (dealloc_ptr == write_ptr) {
+		// The queue is empty
+		return false;
+	}
+
+	uint32_t size = *(uint32_t *)&command_mem[dealloc_ptr];
+
+	if (size == 0) {
+		// End of command buffer wrap down
+		dealloc_ptr = 0;
+		goto tryagain;
+	}
+
+	if (size & 1) {
+		// Still used, nothing can be deallocated
+		return false;
+	}
+
+	dealloc_ptr += (size >> 1) + sizeof(uint32_t);
+	return true;
+}
+
```

Comment:
Because in theory more than on thread could read and execute commands and because these commands can end in a different order w.r.t. the order they had been pushed in and read from the queue, we can't just deallocate them after execution (there could be commands still executing BEFORE that one and we can't advance the read_ptr) so I now have a read_ptr and a dealloc_ptr.

The first one tells where the next command is that should be read, the second one tells where the first possible command to deallocate is. The first bit of the control data now is used to mark usage. We can deallocate the next block as soon as it is not used, and we try to deallocate as many consecutive blocks as we can if we ran out of memory.